FIAT/hellan_herrmann_johnson.py

```diff
@@ -3,79 +3,100 @@
-    """Degrees of freedom for Hellan-Herrmann-Johnson elements."""
-    def __init__(self, ref_el, degree):
+    def __init__(self, ref_el, degree, variant, qdegree):
         sd = ref_el.get_spatial_dimension()
-        if sd != 2:
-            raise ValueError("Hellan-Herrmann-Johnson elements are only"
-                             "defined in dimension 2.")
-
         top = ref_el.get_topology()
+        n = list(map(ref_el.compute_scaled_normal, sorted(top[sd-1])))
         entity_ids = {dim: {i: [] for i in sorted(top[dim])} for dim in sorted(top)}
         nodes = []
 
-        # no vertex dofs
-        edge_dofs, entity_ids[1] = self._generate_edge_dofs(ref_el, degree, 0)
-        nodes.extend(edge_dofs)
-        cell_nodes, entity_ids[sd] = self._generate_cell_dofs(ref_el, degree, len(nodes))
-        nodes.extend(cell_nodes)
+        # Face dofs
+        if variant == "point":
+            # n^T u n evaluated on a Pk lattice
+            for f in sorted(top[sd-1]):
+                cur = len(nodes)
+                pts = ref_el.make_points(sd-1, f, degree + sd)
+                nodes.extend(PointwiseInnerProductEvaluation(ref_el, n[f], n[f], pt)
+                             for pt in pts)
+                entity_ids[sd-1][f].extend(range(cur, len(nodes)))
 
-        super().__init__(nodes, ref_el, entity_ids)
+        elif variant == "integral":
+            # n^T u n integrated against a basis for Pk
+            facet = ref_el.construct_subelement(sd-1)
+            Q = create_quadrature(facet, qdegree + degree)
+            P = polynomial_set.ONPolynomialSet(facet, degree)
+            phis = P.tabulate(Q.get_points())[(0,)*(sd-1)]
+            for f in sorted(top[sd-1]):
+                cur = len(nodes)
+                Q_mapped = FacetQuadratureRule(ref_el, sd-1, f, Q)
+                detJ = Q_mapped.jacobian_determinant()
+                nodes.extend(BidirectionalMoment(ref_el, n[f], n[f]/detJ, Q_mapped, phi) for phi in phis)
+                entity_ids[sd-1][f].extend(range(cur, len(nodes)))
 
-    @staticmethod
-    def _generate_edge_dofs(ref_el, degree, offset):
-        """Generate dofs on edges.
-        On each edge, let n be its normal. For degree=r, the scalar function
-              n^T u n
-        is evaluated at enough points to control P(r).
-        """
-        dim = 1
-        top = ref_el.get_topology()
-        nodes = []
-        entity_ids = {}
-        for entity_id in sorted(top[dim]):
-            pts = ref_el.make_points(dim, entity_id, degree + 2)
-            normal = ref_el.compute_scaled_normal(entity_id)
-            nodes.extend(InnerProduct(ref_el, normal, normal, pt) for pt in pts)
-            num_new_nodes = len(pts)
-            entity_ids[entity_id] = list(range(offset, offset + num_new_nodes))
-            offset += num_new_nodes
-        return nodes, entity_ids
+        # Interior dofs
+        cur = len(nodes)
+        if sd == 2 and variant == "point":
```

Comment:
Do we want to catch the case where `variant` is other than `"point"` or `"integral"` as an exception?

Reply:
that's handled by `check_format_variant`